### PR TITLE
Added a menu for when world gen crashes.

### PR DESCRIPTION
**Pull request by Rartrin**
### What is the new feature?
A menu for when world gen crashes which shows you the name of the world gen pass that caused the issue and the exception message. Also includes both an "Open Logs" and a "Back to Main Menu" button. (See #812 )

### Why should this be part of tModLoader?
1. So that the user knows if whether it is simply world gen taking a long time or if it actually crashed.
2. Generally, even when an exception was thrown during world gen, the game is still completely playable. Some larger mod packs have extremely long load times (10-15 minutes) and much higher chances of crashing worldgen, so having a way to restart world gen without restarting the game, can save a lot of time, 

### Are there alternative designs?
My original implementation involved changing the vanilla world loading screen but I changed it to be its own separate screen which means less changes to vanilla. YAY!

### Sample usage for the new feature
Trying to play with a mod pack that takes 10 minutes to load but only 10 seconds to get softlocked because a world gen pass threw an IndexOutOfRangeException.

### ExampleMod updates
N/A

### CODE REVIEW

**Comment by Rartrin** [comment]
This mod will make it crash on pass 15, which is named "CRASH" and throw an exception with the message "Crashed".
[Crash on Pass 15.zip](https://github.com/tModLoader/tModLoader/files/4453756/Crash.on.Pass.15.zip)

**Comment by Chicken-Bones** [comment]
We should consider forcing mods to call a 'keep alive' method in long running passes, allowing the user to kill worldgen if it doesn't look like progress is being made.

If a pass doesn't call the keepalive once per 100ms in dev mode, we can crash the game with a log message and popup

**Comment by Rartrin** [comment]
For me it is always softlocks due to something throwing an exception like a mod not finding a tile it is looking for and it winds up going outside of an array.
That said, I could try to look into it later, though it is somewhat of a separate issue from what I was doing here. Probably the best way to do it would be make the keep alive be any update to the Progress instance, that way it is somewhat backwards compatible with some mods.
Maybe after 5-10 seconds of no progress updates, pop up a message on the screen saying "mod doesn't seem to be responding" it can show an abort/stop button which take them to the crash screen, otherwise, if an update happens or the pass finishes, it hides the message and button. This way, if a mod actually does work but just doesn't update the progress bar, it will still be usable, while also providing an easy way out if it is actually softlocked. Also, we just tell modders to make sure they update the progress bar, they don't need to use or implement anything new.

File: patches/tModLoader/Terraria.World.Generation/WorldGenerator.cs.patch [hunk 1]
```diff
@@ -13,4 +13,28 @@
...
++				catch(Exception e)
++				{
++					string message = string.Join
++					(
++						"\n",
++						Localization.Language.GetTextValue("tModLoader.WorldGenError"),
++						current2.Name,
++						e.Message
++					);
++					Main.MenuUI.SetState(new ModLoader.UI.UIWorldLoadCrash(message));
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria.World.Generation/WorldGenerator.cs.patch:34]: Is this threadsafe? I do much prefer constructing UI's each time they're used, but it's not the Terraria paradigm.

**Comment by Chicken-Bones** [comment]
Is there a reason you can't use UIErrorMessage?

File: patches/tModLoader/Terraria.World.Generation/WorldGenerator.cs.patch (hunk 1 above)
Comment (Jofairden) [reply patches/tModLoader/Terraria.World.Generation/WorldGenerator.cs.patch:34]: Shouldn't it be setting Main.menuMode to a custom ID? That is how we do this for all menus. Check Interface.cs
( I do want to rewrite this, still )

**Comment by Rartrin** [comment]
Okay, I brought it up to date with the master and converted it over to use UIErrorMessage. I also noticed that the Open Logs button wasn't working. I found the issue and included it. It is a one line fix is all.

**Pull request merged**